File: src/ingestion/document-parsers/directDocumentParser.ts
import { JSONObject } from "../../common/jsonTypes";
import {
  RawDocument,
  IngestedDocument,
  DocumentFragment,
} from "../../document/document";
import { BaseDocumentParser } from "./documentParser";
import { v4 as uuid } from "uuid";
import { Md5 } from "ts-md5";
import {
  CallbackManager,
  ParseNextErrorEvent,
  ParseSuccessEvent,
} from "../../utils/callbacks";

/**
 * Parse a RawDocument directly into a Document, with each DocumentFragment
 * representing a RawDocumentChunk.
 */
export class DirectDocumentParser extends BaseDocumentParser {
  constructor(
    attributes?: JSONObject,
    metadata?: JSONObject,
    callbackManager?: CallbackManager
  ) {
    super(attributes, metadata, callbackManager);
  }

  async parse(rawDocument: RawDocument): Promise<IngestedDocument> {
    const chunks = await rawDocument.getChunkedContent();
    const documentId = uuid();

    const fragments: DocumentFragment[] = [];

    for (const [idx, chunk] of chunks.entries()) {
      const currentFragment: DocumentFragment = {
        fragmentId: uuid(),
        fragmentType: "text",
        documentId,
        metadata: chunk.metadata,
        attributes: {},
        // TODO: figure out blobId
        hash: Md5.hashStr(chunk.content),
        getContent: async () => chunk.content,
        serialize: async () => JSON.stringify(chunk),
        previousFragment: idx > 0 ? fragments[idx - 1] : undefined,
      };

      if (idx > 0) {
        fragments[idx - 1].nextFragment = currentFragment;
      }

      fragments.push(currentFragment);
    }

    const out = {
      documentId,
      rawDocument,
      collectionId: rawDocument.collectionId,
      fragments,
      serialize: () => this.serialize(rawDocument),
      attributes: this.attributes,
      metadata: this.metadata,
    };
    const event: ParseSuccessEvent = {
      name: "onParseSuccess",
      ingestedDocument: out,
    };
    this.callbackManager?.runCallbacks(event);
    return out;
  }

  // TODO: This is mainly useful for parsing the raw document as a stream/buffer
  // instead of chunking the full document at once when loading.
  // Revisit when we add support for that
  async parseNext(
    _rawDocument: RawDocument,
    _previousFragment?: DocumentFragment,
    _take?: number
  ): Promise<DocumentFragment> {
    const err = new Error("Method not implemented.");
    const event: ParseNextErrorEvent = {
      name: "onParseNextError",
      error: err,
    };
    await this.callbackManager?.runCallbacks(event);
    throw err;
  }
}

File: python/src/semantic_retrieval/document_parsers/direct_document_parser.py
import uuid
from typing import Dict, List, Optional

from result import Err, Ok, Result
from semantic_retrieval.document.document import (
    DirectDocumentFragment,
    DocumentFragment,
    DocumentFragmentType,
    IngestedDocument,
    RawDocument,
)
from semantic_retrieval.ingestion.document_parsers.document_parser import (
    BaseDocumentParser,
)
from semantic_retrieval.utils.callbacks import CallbackManager, Traceable


class DirectDocumentParser(BaseDocumentParser, Traceable):
    def __init__(
        self,
        attributes: Dict[str, str],
        metadata: Optional[Dict[str, str]],
        callback_manager: Optional[CallbackManager] = None,
    ):
        super().__init__(attributes, metadata or {}, callback_manager)

    async def parse(
        self, raw_document: RawDocument
    ) -> Result[IngestedDocument, str]:
        # Not using chunked content
        content = (await raw_document.get_content()).unwrap()

        content_result = await raw_document.get_content()

        # TODO callbacks

        match content_result:
            case Ok(content):
                document_id = raw_document.document_id

                # Only make one fragment with the entire contents of the document for now
                fragments: List[DirectDocumentFragment] = [
                    DirectDocumentFragment(
                        fragment_id=str(uuid.uuid4()),
                        fragment_type=DocumentFragmentType.TEXT,
                        document_id=document_id,
                        content=content,
                        hash=None,
                        metadata={},
                        attributes={},
                    )
                ]

                return Ok(
                    IngestedDocument(
                        raw_document=raw_document,
                        document_id=document_id,
                        collection_id=raw_document.collection_id,
                        fragments=fragments,
                        metadata={},
                        attributes={},
                    )
                )
            case Err(err):
                return Err(err)

    def parse_next(
        self,
        raw_document: RawDocument,
        previous_fragment: Optional[DocumentFragment] = None,
        take: Optional[int] = None,
    ) -> DocumentFragment:
        raise NotImplementedError()
